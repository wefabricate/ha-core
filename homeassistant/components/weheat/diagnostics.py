"""Diagnostics support for Weheat."""

from __future__ import annotations

from typing import Any

from homeassistant.components.diagnostics import async_redact_data
from homeassistant.const import CONF_ACCESS_TOKEN
from homeassistant.core import HomeAssistant

from . import WeheatConfigEntry
from .const import CONF_REFRESH_TOKEN

TO_REDACT = {CONF_ACCESS_TOKEN, CONF_REFRESH_TOKEN}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: WeheatConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""

    diag_data = {}

    for coordinator in entry.runtime_data:
        diag_data[coordinator.heatpump_id] = {
            "heat_pump_info": coordinator.heat_pump_info.__dict__,
            "heat_pump_data": coordinator.data.raw_content,
            "session_valid": coordinator.session.valid_token,
            "session_token": async_redact_data(coordinator.session.token, TO_REDACT),
        }

    return diag_data
