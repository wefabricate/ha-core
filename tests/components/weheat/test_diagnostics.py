"""Tests for the diagnostics data provided by the Withings integration."""

from unittest.mock import AsyncMock

import pytest
from syrupy import SnapshotAssertion
from syrupy.filters import props
from weheat.abstractions.discovery import HeatPumpDiscovery

from homeassistant.core import HomeAssistant

from . import setup_integration

from tests.common import MockConfigEntry
from tests.components.diagnostics import get_diagnostics_for_config_entry
from tests.typing import ClientSessionGenerator


@pytest.mark.usefixtures("setup_credentials")
async def test_diagnostics_config_entry(
    hass: HomeAssistant,
    hass_client: ClientSessionGenerator,
    mock_weheat_discover: AsyncMock,
    mock_weheat_heat_pump: AsyncMock,
    mock_heat_pump_info: HeatPumpDiscovery.HeatPumpInfo,
    mock_config_entry: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the Weheat setup."""
    await setup_integration(hass, mock_config_entry)

    assert await get_diagnostics_for_config_entry(
        hass, hass_client, mock_config_entry
    ) == snapshot(exclude=props("expires_at", "modified_at", "repr"))
