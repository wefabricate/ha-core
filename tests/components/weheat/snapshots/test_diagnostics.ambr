# serializer version: 1
# name: test_diagnostics_config_entry
  dict({
    '0000-1111-2222-3333': dict({
      'heat_pump_data': dict({
        'dummy': 'test',
      }),
      'heat_pump_info': dict({
        'has_dhw': True,
        'model': 'Test Model',
        'name': None,
        'sn': 'SN-Test-This',
        'uuid': '0000-1111-2222-3333',
      }),
      'session_token': dict({
        'access_token': '**REDACTED**',
        'expires_in': 60,
        'refresh_token': '**REDACTED**',
        'type': 'Bearer',
      }),
      'session_valid': True,
    }),
  })
# ---
